File: argon-plugin-js-api/shared/src/main/scala/dev/argon/plugin/jsapi/Options.scala
package dev.argon.plugin.jsapi

import dev.argon.plugin.jsapi.proto.Toml
import scalajs.js

sealed trait OptionDecodeResult[A] extends js.Any
trait OptionDecodeResultValue[A] extends OptionDecodeResult[A] {
  val value: A
}

trait OptionDecodeResultError[A] extends OptionDecodeResult[A] {
  val errorMessage: String
}

trait OptionDecoder[Options] extends js.Any {
  def decode(resourceFactory: ResourceFactory, value: Toml): OptionDecodeResult[Options]
  def defaultValue: Options | Null
}

trait OptionCodec[Options] extends OptionDecoder[Options] {
  def encode(recorder: ResourceRecorder, value: Options): js.Promise[Toml]
  def skipForField(output: Options): Boolean
}

trait OutputInfo[Output] extends js.Any {
  def getValue(output: Output): FileSystemResource
}

trait OutputHandler[Output] extends js.Any {
  def options: js.Map[js.Array[String], OutputInfo[Output]]
}

File: plugin/api/js/src/options.ts
import type {ResourceFactory, ResourceRecorder, FileSystemResource} from "./resource.js";
import type {Toml} from "./proto/util.js";

export type OptionDecodeResult<A> = OptionDecodeSuccess<A> | OptionDecodeFailure;

export interface OptionDecodeSuccess<A> {
    readonly ok: true;
    readonly value: A;
}

export interface OptionDecodeFailure {
    readonly ok: false;
    readonly errorMessage: string;
}

export interface OptionDecoder<Options> {
    decode(resourceFactory: ResourceFactory, value: Toml): OptionDecodeResult<Options>;
}

export interface OptionCodec<Options> extends OptionDecoder<Options> {
    encode(recorder: ResourceRecorder, value: Options): Promise<Toml>;
}

export interface OutputInfo<Output> {
    getValue(output: Output): FileSystemResource;
}

export interface OutputHandler<Output> {
    get options(): Map<string, OutputInfo<Output>>;
}


